File: backend/api/history_api.py
"""History API — manage past sessions, messages, and exports."""

from __future__ import annotations

import logging

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from backend.services.history_service import session_history_service

logger = logging.getLogger("tenderclaw.api.history")
router = APIRouter(tags=["history"])


class SessionSummary(BaseModel):
    session_id: str
    title: str
    created_at: str
    message_count: int
    model: str
    preview: str
    total_cost_usd: float


class SessionDetail(BaseModel):
    session_id: str
    title: str
    created_at: str
    updated_at: str
    model: str
    message_count: int
    messages: list[dict]
    total_usage: dict
    total_cost_usd: float
    working_directory: str


class HistoryListResponse(BaseModel):
    sessions: list[SessionSummary]
    total: int


class HistoryPageResponse(BaseModel):
    entries: list[dict]
    total: int
    has_more: bool
    cursor: str | None


class MessagePageResponse(BaseModel):
    messages: list[dict]
    has_more: bool
    cursor: str | None


class DeleteResponse(BaseModel):
    status: str
    message: str


@router.get("", response_model=HistoryPageResponse)
async def get_sessions(
    limit: int = Query(20, ge=1, le=100),
    before_id: str | None = None,
    search: str | None = None,
) -> HistoryPageResponse:
    """Get paginated session history."""
    return session_history_service.get_sessions(
        limit=limit,
        before_id=before_id,
        search=search,
    )


@router.get("/legacy", response_model=HistoryListResponse)
async def list_sessions(
    limit: int = 50,
    offset: int = 0,
    keyword: str | None = None,
    date_from: str | None = None,
    date_to: str | None = None,
) -> HistoryListResponse:
    """List past sessions with optional filtering (legacy)."""
    sessions = session_history_service.list_sessions(
        limit=limit,
        offset=offset,
        keyword=keyword,
        date_from=date_from,
        date_to=date_to,
    )
    return HistoryListResponse(sessions=sessions, total=len(sessions))


@router.get("/{session_id}", response_model=SessionDetail)
async def get_session(session_id: str) -> SessionDetail:
    """Get full session details including messages."""
    detail = session_history_service.get_session_detail(session_id)
    if not detail:
        raise HTTPException(status_code=404, detail=f"Session not found: {session_id}")
    return SessionDetail(**detail)


@router.delete("/{session_id}", response_model=DeleteResponse)
async def delete_session(session_id: str) -> DeleteResponse:
    """Delete a session from history."""
    success = session_history_service.delete_session(session_id)
    if not success:
        raise HTTPException(status_code=404, detail=f"Session not found: {session_id}")
    return DeleteResponse(status="ok", message=f"Session {session_id} deleted")


@router.get("/{session_id}/messages", response_model=MessagePageResponse)
async def get_session_messages(
    session_id: str,
    limit: int = Query(50, ge=1, le=200),
    before_id: str | None = None,
) -> MessagePageResponse:
    """Get paginated messages for a session."""
    result = session_history_service.get_session_messages_paginated(
        session_id,
        limit=limit,
        before_id=before_id,
    )
    if result is None:
        raise HTTPException(status_code=404, detail=f"Session not found: {session_id}")
    return MessagePageResponse(**result)


@router.get("/{session_id}/messages/all")
async def get_session_messages_all(session_id: str) -> list[dict]:
    """Get all messages for a specific session (legacy)."""
    messages = session_history_service.get_session_messages(session_id)
    if messages is None:
        raise HTTPException(status_code=404, detail=f"Session not found: {session_id}")
    return messages


@router.post("/export/{session_id}")
async def export_session(session_id: str) -> dict:
    """Export a session to JSON format."""
    export_data = session_history_service.export_session(session_id)
    if not export_data:
        raise HTTPException(status_code=404, detail=f"Session not found: {session_id}")
    return export_data


@router.post("/import")
async def import_session(data: dict) -> dict:
    """Import a session from exported JSON."""
    session_id = session_history_service.import_session(data)
    if not session_id:
        raise HTTPException(status_code=400, detail="Invalid session data")
    return {"status": "ok", "session_id": session_id}


@router.post("/export-all")
async def export_all_sessions() -> dict:
    """Export all sessions to JSON."""
    return session_history_service.export_all_sessions()

File: frontend/src/api/historyApi.ts
import { SessionSummary, SessionDetail } from "../types/history";

const API_BASE = "/api/history";

export interface HistoryEntry {
  session_id: string;
  title: string;
  created_at: string;
  message_count: number;
  last_message: string;
  total_cost_usd: number;
  model: string;
}

export interface HistoryPage {
  entries: HistoryEntry[];
  total: number;
  has_more: boolean;
  cursor: string | null;
}

export interface MessagePage {
  messages: Message[];
  has_more: boolean;
  cursor: string | null;
}

export interface Message {
  role: string;
  content: string | unknown[];
  message_id?: string;
  timestamp?: string;
}

export async function listSessions(params?: {
  limit?: number;
  offset?: number;
  keyword?: string;
  dateFrom?: string;
  dateTo?: string;
}): Promise<{ sessions: SessionSummary[]; total: number }> {
  const searchParams = new URLSearchParams();
  if (params?.limit) searchParams.set("limit", String(params.limit));
  if (params?.offset) searchParams.set("offset", String(params.offset));
  if (params?.keyword) searchParams.set("keyword", params.keyword);
  if (params?.dateFrom) searchParams.set("date_from", params.dateFrom);
  if (params?.dateTo) searchParams.set("date_to", params.dateTo);

  const res = await fetch(`${API_BASE}/legacy?${searchParams}`);
  if (!res.ok) throw new Error("Failed to fetch sessions");
  return res.json();
}

export async function getHistory(params: {
  limit?: number;
  before_id?: string;
  search?: string;
}): Promise<HistoryPage> {
  const searchParams = new URLSearchParams();
  if (params.limit) searchParams.set("limit", String(params.limit));
  if (params.before_id) searchParams.set("before_id", params.before_id);
  if (params.search) searchParams.set("search", params.search);

  const res = await fetch(`${API_BASE}?${searchParams}`);
  if (!res.ok) throw new Error("Failed to fetch history");
  return res.json();
}

export async function getSessionDetail(sessionId: string): Promise<SessionDetail> {
  const res = await fetch(`${API_BASE}/${sessionId}`);
  if (!res.ok) throw new Error("Failed to fetch session detail");
  return res.json();
}

export async function getSession(sessionId: string): Promise<SessionDetail> {
  const res = await fetch(`${API_BASE}/${sessionId}`);
  if (!res.ok) throw new Error("Failed to fetch session");
  return res.json();
}

export async function getSessionMessages(sessionId: string): Promise<Message[]> {
  const res = await fetch(`${API_BASE}/${sessionId}/messages/all`);
  if (!res.ok) throw new Error("Failed to fetch messages");
  return res.json();
}

export async function getMessages(sessionId: string, params?: {
  limit?: number;
  before_id?: string;
}): Promise<MessagePage> {
  const searchParams = new URLSearchParams();
  if (params?.limit) searchParams.set("limit", String(params.limit));
  if (params?.before_id) searchParams.set("before_id", params.before_id);

  const res = await fetch(`${API_BASE}/${sessionId}/messages?${searchParams}`);
  if (!res.ok) throw new Error("Failed to fetch messages");
  return res.json();
}

export async function deleteSession(sessionId: string): Promise<void> {
  const res = await fetch(`${API_BASE}/${sessionId}`, { method: "DELETE" });
  if (!res.ok) throw new Error("Failed to delete session");
}

export async function exportSession(sessionId: string): Promise<unknown> {
  const res = await fetch(`${API_BASE}/export/${sessionId}`, { method: "POST" });
  if (!res.ok) throw new Error("Failed to export session");
  return res.json();
}

export async function exportAllSessions(): Promise<unknown> {
  const res = await fetch(`${API_BASE}/export-all`, { method: "POST" });
  if (!res.ok) throw new Error("Failed to export all sessions");
  return res.json();
}

export async function importSession(data: unknown): Promise<{ session_id: string }> {
  const res = await fetch(`${API_BASE}/import`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error("Failed to import session");
  return res.json();
}
